### PR TITLE
feat: add crates for settings plugins

### PR DESCRIPTION
**Pull request by bcressey**
**Issue #, if available:**
N/A

**Description of changes:**
Settings plugins are an intermediate step toward the planned world of settings extensions. Variants can define a monolithic settings struct that combines custom and shared types, which can be loaded at runtime by Bottlerocket's existing API server and datastore implementations.

Since Rust lacks a stable ABI, the key to making this work is for the host program (the API server or datastore creator) to load the plugin as a cdylib that implements a particular FFI contract. The abi_stable crate does the heavy lifting here with its support for load-time type checking and FFI-safe wrapper types.

Plugins must define a struct and implement serialize, deserialize, and default methods for that struct. That allows the host program to instantiate a new version of the struct, make increment changes, save the result to the datastore, and load it again, all without compile- time access to the type. That in turn means the host programs can be built once and support arbitrary settings at runtime.

**Terms of contribution:**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: bottlerocket-settings-derive/src/lib.rs
Comment (sam-berning) [inline bottlerocket-settings-derive/src/lib.rs]: nit: this looks like it's wrapped into the `sabi_extern_fn` macro: https://docs.rs/abi_stable/latest/abi_stable/attr.sabi_extern_fn.html

**Review by sam-berning — APPROVED** [review]

**Comment by bcressey** [comment]
⬆️ force push splits the commit to better attribute the `Serialize` and `Deserialize` work to @sam-berning.

File: bottlerocket-settings-plugin/src/settings.rs
Comment (webern) [inline bottlerocket-settings-plugin/src/settings.rs]: Why not do the error unwrap/panic here?

**Comment by bcressey** [comment]
⬆️ force push removes the redundant `abi_stable::extern_fn_panic_handling!` call.

**Review by webern — APPROVED** [review]

**Review by sam-berning — APPROVED** [review]

**Comment by bcressey** [comment]
⬆️ force push consolidates the error handling when loading the plugin

**Review by webern — APPROVED** [review]

**Comment by bcressey** [comment]
⬆️ force push rebase to fix `Cargo.toml` conflict - no other changes

**Comment by bcressey** [comment]
⬆️ adds back the cargo deny exception for syn v1 - that didn't stay away for very long!

**Review by webern — APPROVED** [review]

File: bottlerocket-settings-plugin/src/settings.rs
Comment (cbgbt) [inline bottlerocket-settings-plugin/src/settings.rs]: I'm not sure I follow why this can't just be `serde_json::from_value<BottlerocketSettingsProvider>(json_value)`?

I'm assuming it has something to do with how the host/plugin interact with this interface.

Comment (cbgbt) [reply bottlerocket-settings-plugin/src/settings.rs]: I chatted a bit with @sam-berning and I *believe* that once we've done the initial deserialization with `JsonValue::deserialize`, that the host should have all of the information needed to do a direct conversion of this value to type `T` via `serde_json::from_value`. Likewise, we can do the same in the `Serialize` impl.

There may still be a gap in my understanding here, though.

Comment (bcressey) [reply bottlerocket-settings-plugin/src/settings.rs]: Nice catch. Looks like it works for the `apiserver` tests, going to try it in a real variant.

Comment (bcressey) [reply bottlerocket-settings-plugin/src/settings.rs]: Spoke too soon, I end up with errors like these:
```
❯ cargo test
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.28s
     Running unittests src/lib.rs (/home/fedora/brdev/sources/target/debug/deps/apiserver-dabc4cdb2227b421)

running 10 tests
test server::controller::test::get_metadata_all_works ... ok
test server::controller::test::get_metadata_keys_works ... ok
test server::controller::test::get_services_names_works ... ok
test server::controller::test::commit_works ... FAILED
test server::controller::test::get_configuration_files_prefix_works ... ok
test server::controller::test::get_settings_works ... FAILED
test server::controller::test::set_settings_works ... FAILED
test server::controller::test::get_settings_prefix_works ... FAILED
test server::controller::test::get_services_prefix_works ... ok
test server::controller::test::get_settings_keys_works ... FAILED

failures:

---- server::controller::test::commit_works stdout ----
thread 'server::controller::test::commit_works' panicked at api/apiserver/src/server/controller.rs:854:52:
called `Result::unwrap()` on an `Err` value: Deserialization { given: "settings.", source: Message { msg: "invalid type: string \"{\\\"motd\\\":\\\"json string\\\"}\", expected raw value" } }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- server::controller::test::get_settings_works stdout ----
thread 'server::controller::test::get_settings_works' panicked at api/apiserver/src/server/controller.rs:601:60:
called `Result::unwrap()` on an `Err` value: Deserialization { given: "settings.", source: Message { msg: "invalid type: string \"{\\\"motd\\\":\\\"json string\\\"}\", expected raw value" } }

---- server::controller::test::set_settings_works stdout ----
thread 'server::controller::test::set_settings_works' panicked at api/apiserver/src/server/controller.rs:777:86:
called `Result::unwrap()` on an `Err` value: Error("invalid type: string \"{\\\"motd\\\":\\\"tz\\\"}\", expected raw value", line: 0, column: 0)

---- server::controller::test::get_settings_prefix_works stdout ----
thread 'server::controller::test::get_settings_prefix_works' panicked at api/apiserver/src/server/controller.rs:618:14:
called `Result::unwrap()` on an `Err` value: Deserialization { given: "settings.", source: Message { msg: "invalid type: string \"{\\\"motd\\\":\\\"json string\\\"}\", expected raw value" } }

---- server::controller::test::get_settings_keys_works stdout ----
thread 'server::controller::test::get_settings_keys_works' panicked at api/apiserver/src/server/controller.rs:658:82:
called `Result::unwrap()` on an `Err` value: Deserialization { given: "given keys", source: Message { msg: "invalid type: string \"{\\\"motd\\\":\\\"json string 1\\\"}\", expected raw value" } }
```

Comment (sam-berning) [reply bottlerocket-settings-plugin/src/settings.rs]: ugh, I was a little worried that this wouldn't work on Deserialization but I wasn't sure why -- but I think that DynTrait's Deserialize specifically needs to be done on a string, or at least that's what this section seems to imply: https://docs.rs/abi_stable/latest/abi_stable/struct.DynTrait.html#trait-object

**Comment by bcressey** [comment]
⬆️ force push reverts to the working `Deserialize` impl

**Review by cbgbt — APPROVED** [review]
With the tests passing, I don't want my understanding of the serde bits holding up this work. I'll look into it more after it is merged.

**Pull request merged**
